src/types/acls.rs

```diff
@@ -0,0 +1,351 @@
+use crate::error::{Error, ParseError};
+use crate::parse::try_handle_unilateral;
+use crate::types::UnsolicitedResponse;
+use imap_proto::types::AclRight;
+use imap_proto::Response;
+use ouroboros::self_referencing;
+use std::borrow::Cow;
+use std::collections::HashSet;
+use std::fmt::{Display, Formatter};
+use std::sync::mpsc;
+
+/// enum used for set_acl to specify how the ACL is to be modified.
+pub enum AclModifyMode {
+    /// Replace all ACLs on the identifier for the mailbox
+    Replace,
+    /// Add the given ACLs to the identifier for the mailbox
+    Add,
+    /// Remove the given ACLs from the identifier for the mailbox
+    Remove,
+}
+
+/// Helpful wrapper around the ACL rights vector
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclRightList {
+    pub(crate) data: HashSet<AclRight>,
+}
+
+impl AclRightList {
+    /// Returns if the AclRightList has the provided ACL (either as a char or an AclRight enum)
+    pub fn has_right<T: Into<AclRight>>(&self, right: T) -> bool {
+        self.data.contains(&right.into())
+    }
+}
+
+impl Display for AclRightList {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut v: Vec<char> = self.data.iter().map(|c| char::from(*c)).collect();
+
+        v.sort_unstable();
+
+        write!(f, "{}", v.into_iter().collect::<String>())
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRightList {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRightList {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRightList {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl From<&str> for AclRightList {
+    fn from(i: &str) -> Self {
+        i.chars()
+            .into_iter()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into()
+    }
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the getacl IMAP command
+#[self_referencing]
+pub struct Acl {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
```

Comment:
I think this would become nicer if you kept the mailbox in this struct. So:
```suggestion
    #[borrows(data)]
    #[covariant]
    pub(crate) mailbox: Cow<'this, str>,
    /// The list of identifier/rights pairs for the mailbox
    #[borrows(data)]
    #[covariant]
    pub(crate) acls: Vec<AclEntry<'a>>,
```

It'll also let you get rid of the `InnerAcl` type.

Reply:
I started that way, however the way ouroboros's build works it has separate parsers for each member. (mailbox_builder, and acls_builder both receicing the [u8] input). so I couldn't find any easy way to NOT end up parsing twice (which definitely would be undesirable). 

What I did instead was replace the named inner structs with inline anonymous tuples.

Reply:
Ahh, that's awkward. I think I'm in favor of reverting 00d3b288f2476b0dd87942815bbfb794cef43c00 then — tuple types with 4+ members get unwieldy and error-prone real fast!

Reply:
I reverted that commit.